File: Codes/1090-largest-values-from-labels.py
'''
Author: LetMeFly
Date: 2023-05-23 11:46:48
LastEditors: LetMeFly
LastEditTime: 2023-05-23 11:54:00
'''
from typing import List
from collections import defaultdict


class Solution:
    def largestValsFromLabels(self, values: List[int], labels: List[int], numWanted: int, useLimit: int) -> int:
        v = list(zip(values, labels))
        v.sort(key=lambda x : -x[0])
        ans = 0
        cnt = 0
        ma = defaultdict(int)
        for value, label in v:
            if ma[label] == useLimit:
                continue
            ma[label] += 1
            cnt += 1
            ans += value
            if cnt == numWanted:
                break
        return ans

File: Codes/1090-largest-values-from-labels.cpp
/*
 * @Author: LetMeFly
 * @Date: 2023-05-23 11:46:42
 * @LastEditors: LetMeFly
 * @LastEditTime: 2023-05-23 11:50:41
 */
#ifdef _WIN32
#include "_[1,2]toVector.h"
#endif

class Solution {
public:
    int largestValsFromLabels(vector<int>& values, vector<int>& labels, int numWanted, int useLimit) {
        vector<pair<int, int>> v(values.size());
        for (int i = 0; i < values.size(); i++) {
            v[i] = {values[i], labels[i]};
        }
        sort(v.begin(), v.end(), [&](const pair<int, int>& a, const pair<int, int>& b) {
            return a.first > b.first;
        });
        unordered_map<int, int> ma;
        int cnt = 0;
        int ans = 0;
        for (auto&& [value, label] : v) {
            if (ma[label] == useLimit) {
                continue;
            }
            ma[label]++;
            cnt++;
            ans += value;
            if (cnt == numWanted) {
                break;
            }
        }
        return ans;
    }
};